toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1092,6 +1106,91 @@ func (g *PkgGraph) AddGoalNode(goalName string, packages []*pkgjson.PackageVer,
 		}
 	}
 
+	// Expand the goal node if requested
+	if extraLayers > 0 {
+		g.addGoalNodeLayers(goalNode, extraLayers)
+	}
+
+	return
+}
+
+// addGoalNodeLayers will expand a goal node by one layer. For example, if the goaled node is "foo" (ie the goal node
+// points to "foo") and extraLevels is 1, the goal node will now link to all nodes which depend on "foo" as well as
+// "foo" itself. A node is considered to depend on "foo" if it is a run node that has edges connecting it to "foo"
+// without any other run nodes in between.
+//
+//	i.e., if "foo_run" -> "foo_build" -> "<Some Meta Node>" -> "bar_run" -> "bar_build", and we are expanding from "bar"
+//	with layers=1, only "foo_run" will be added to the goal nodes since "foo_build" and "<Some Meta Node>" are not run nodes.
+func (g *PkgGraph) addGoalNodeLayers(goalNode *PkgNode, layers int) {
+	logger.Log.Debugf("Expanding goal node %s by %d layers", goalNode.GoalName, layers)
+
+	var expandedGoalNodes []*PkgNode
+	// Use a set to keep track of the nodes we already added so we can avoid processing them again
+	expandedGoalNodesSet := make(map[*PkgNode]bool)
+
+	// Start with the already selected nodes which make up the goal.
+	initialGoalNodes := []*PkgNode{}
+	for _, selectedNode := range graph.NodesOf(g.From(goalNode.ID())) {
+		initialGoalNodes = append(initialGoalNodes, selectedNode.(*PkgNode))
+	}
+
+	// For each node in the current layer, add all of the nodes that depend on it, then repeat as many times as requested
+	expandedGoalNodes = initialGoalNodes
+	for i := 0; i < layers; i++ {
+		expandedGoalNodes = append(expandedGoalNodes, g.getNextGoalLayer(expandedGoalNodesSet, expandedGoalNodes)...)
+	}
+
+	// Add the new edges if they are missing
+	for _, n := range expandedGoalNodes {
+		if n == goalNode {
```

Comment:
Should we also check that `n` is not inside `initialGoalNodes`? I think it won't matter for the build itself, but the graph will be a bit easier to read.

Reply:
It should be same to assume that all nodes in `initialGoaldNodes` DO have an edge to `g` already, which means they should be skipped.

Reply:
Ah, I see, I initially misunderstood. Now a reverse question came to my mind: if my understanding is correct, all true goal nodes (of type `TypeGoal`) are never in a cycle. And `goalNode` is a true goal node, so do we even need this check?

Reply:
Its to make sure we don't add a self-edge from goal -> goal. When we get the list of all dependents of the goal'd nodes, we will get the goal node itself back again. No reason to add an edge from the goal to the goal.

I'll add a comment to this affect.

Reply:
I'm still going to leave it in with an eye to another PR which will allow arbitrary goal node creation from a list of other nodes as a defensive check.